### PR TITLE
[#1193] improvement(docs): Add the document about how to debug trino connector locally.

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Add a document about how to debug develop and debug `trino-connector` locally. 

### Why are the changes needed?

To make users more knowledgeable about Gravitino trino connector.

Fix: #1193 

### Does this PR introduce _any_ user-facing change?

N/A.

### How was this patch tested?

N/A.

### CODE REVIEW

File: docs/assets/trino/pom.xml [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+            <groupId>com.datastrato.gravitino</groupId>
+            <artifactId>bundled-catalog</artifactId>
+            <version>0.4.0-SNAPSHOT</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.datastrato.gravitino</groupId>
+            <artifactId>catalog-common</artifactId>
+            <version>0.4.0-SNAPSHOT</version>
+        </dependency>
```
Comment (xiacongling) [inline docs/assets/trino/pom.xml:162]: these packages are not in maven central repository. can you publish them? or can we build and publish locally them from source code?

Comment (xiacongling) [reply docs/assets/trino/pom.xml:162]: shall we use current snapshots (0.5.0-SNAPSHOT)?

Comment (yuqi1129) [reply docs/assets/trino/pom.xml:162]: I believe you can install them with the latest version 0.5.0-SNAPSHOT locally.

**Comment by yuqi1129** [comment]
@diqiu50 , 
Could you kindly review it for me?

File: docs/assets/trino/pom.xml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        <dependency>
+            <groupId>io.trino</groupId>
+            <artifactId>trino-testing</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>com.datastrato.gravitino</groupId>
+            <artifactId>gravitino-client-java-runtime</artifactId>
```
Comment (xiacongling) [inline docs/assets/trino/pom.xml:148]: there is no such artifact after a local publish. do you mean `client-java-runtime`?

File: docs/assets/trino/pom.xml [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        <dependency>
+            <groupId>com.datastrato.gravitino</groupId>
+            <artifactId>bundled-catalog</artifactId>
+            <version>0.4.0-SNAPSHOT</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.datastrato.gravitino</groupId>
+            <artifactId>catalog-common</artifactId>
```
Comment (xiacongling) [inline docs/assets/trino/pom.xml:160]: no such artifact either. do you mean `common`

File: docs/trino-connector/development.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+```curl
+curl -X POST -H "Content-Type: application/json" -d '{"name":"test","comment":"comment","properties":{}}' http://localhost:8090/api/metalakes
+
+curl -X POST -H "Content-Type: application/json" -d '{"name":"mysql_catalog3","type":"RELATIONAL","comment":"comment","provider":"jdbc-mysql", "properties":{
+  "jdbc-url": "jdbc:mysql://127.0.0.1:3306?useSSL=false&allowPublicKeyRetrieval=true",
+  "jdbc-user": "root",
+  "jdbc-password": "123456",
+  "jdbc-driver": "com.mysql.cj.jdbc.Driver"
+}}' http://localhost:8090/api/metalakes/test/catalogs
+```
```
Comment (xiacongling) [inline docs/trino-connector/development.md:27]: catalogs are innecessary for setting up dev environment.

File: docs/trino-connector/development.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+then you can see the `gravitino-trino-connecor` source files and directories in the `trino-gravitino` module as follows:
+
+![trino-gravitino-structure](../assets/trino/add-link.jpg)
+
+5. Change the `pom.xml` file in the `trino-gravitino` module accordingly. This is a [example](../assets/trino/pom.xml) of the `pom.xml` file in the `trino-gravitino` module.
+6. Try to compile module `trino-gravitino` to see if there are any errors. 
+```shell
+mvn clean -pl 'plugin/trino-gravitino' package -DskipTests -Dcheckstyle.skip -Dair.check.skip-checkstyle=true -DskipTests -Dair.check.skip-all=true
+```
```
Comment (xiacongling) [inline docs/trino-connector/development.md:57]: 1. maven wrapper is more recommanded
2. other libs and plugins need to be build

**Review by xiacongling — CHANGES_REQUESTED** [review]

**Comment by xiacongling** [comment]
Setting up with Trino 421 successfully. I've encountered some problems during the setup:

1. recommanded to use Trino ≥426 (unfortunately, we are using Trino 421). in my case, `ConnectorMetadata.dropSchema` has only 2 arguments (cascade dropping was not supportted yet and this interface method was removed in trino 426), so i need to update gravitino's implementation, rebuild the project and publish artifacts to local maven repository;
2. struggled with signing for some time when publishing my own artifactories to local repository, can we make signing an optional task and skip it by default? we only need to apply signing in public releases, right?

**Comment by yuqi1129** [comment]
@diqiu50 

> recommanded to use Trino ≥426 (unfortunately, we are using Trino 421). in my case, ConnectorMetadata.dropSchema has only 2 arguments (cascade dropping was not supportted yet and this interface method was removed in trino 426), so i need to update gravitino's implementation, rebuild the project and publish artifacts to local maven repository;

About this point, please take a look if we need to modify the supported versions, as we only test and verify versions newer than 426. 

> struggled with signing for some time when publishing my own artifactories to local repository, can we make signing an optional task and skip it by default? we only need to apply signing in public releases, right?

Yeah, it would be great help if we can skip the signing stage. currently, you can use `maven` to install to your local repo, or can you work on it?

**Comment by diqiu50** [comment]
> @diqiu50
> 
> > recommanded to use Trino ≥426 (unfortunately, we are using Trino 421). in my case, ConnectorMetadata.dropSchema has only 2 arguments (cascade dropping was not supportted yet and this interface method was removed in trino 426), so i need to update gravitino's implementation, rebuild the project and publish artifacts to local maven repository;
> 
> About this point, please take a look if we need to modify the supported versions, as we only test and verify versions newer than 426.
> 
> > struggled with signing for some time when publishing my own artifactories to local repository, can we make signing an optional task and skip it by default? we only need to apply signing in public releases, right?
> 
> Yeah, it would be great help if we can skip the signing stage. currently, you can use `maven` to install to your local repo, or can you work on it?

I have verified it; indeed, the dropSchema interface in Trino-421 does not include 'cascade'. We need to update the documentation.

I think we need to conduct a test later to check the compatibility with different Trino versions.

File: docs/trino-connector/development.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+ln -s gravitino/path/to/trino-connector/src/main/java trino/path/to/trino-gravitino/src/main/java
+```
+then you can see the `gravitino-trino-connecor` source files and directories in the `trino-gravitino` module as follows:
+
+![trino-gravitino-structure](../assets/trino/add-link.jpg)
+
+5. Change the `pom.xml` file in the `trino-gravitino` module accordingly. This is a [example](../assets/trino/pom.xml) of the `pom.xml` file in the `trino-gravitino` module.
+6. Try to compile module `trino-gravitino` to see if there are any errors. 
+```shell
+mvn clean -pl 'plugin/trino-gravitino' package -DskipTests -Dcheckstyle.skip -Dair.check.skip-checkstyle=true -DskipTests -Dair.check.skip-all=true
```
Comment (diqiu50) [inline docs/trino-connector/development.md:56]: We need to build the Trino as well

File: docs/assets/trino/pom.xml (hunk 2 above)
Comment (diqiu50) [reply docs/assets/trino/pom.xml:148]: I suggest providing a script to install Graviton-related dependencies into the local Maven repository.

File: docs/trino-connector/development.md (hunk 6 above)
Comment (diqiu50) [reply docs/trino-connector/development.md:56]: I have an error:
```
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.trino:trino-maven-plugin:12:generate-service-descriptor (default-generate-service-descriptor) on project trino-graviton:
[ERROR]
[ERROR] Existing service descriptor for io.trino.spi.Plugin found in output directory.
[ERROR] -> [Help 1]
```

The solution is remove the file `trino-connector/src/main/resources/META-INF/services/io.trino.spi.Plugin`
Do you have this problem

Comment (xiacongling) [reply docs/trino-connector/development.md:56]: you can install trino modules (exluding gravitino) at first, using:
`mvn -pl '!plugin/trino-gravitino' clean install -DskipTests`

File: docs/assets/trino/pom.xml (hunk 2 above)
Comment (yuqi1129) [reply docs/assets/trino/pom.xml:148]: > there is no such artifact after a local publish. do you mean `client-java-runtime`?

Exactly.

Comment (yuqi1129) [reply docs/assets/trino/pom.xml:148]: > I suggest providing a script to install Graviton-related dependencies into the local Maven repository.

I will convert the artifacts to the standard release versions and the user can download them from Maven central repo.

File: docs/trino-connector/development.md (hunk 4 above)
Comment (yuqi1129) [reply docs/trino-connector/development.md:27]: Can you use the Trino command line to display the Gravition catalog if you do not create catalogs here?

File: docs/trino-connector/development.md (hunk 5 above)
Comment (yuqi1129) [reply docs/trino-connector/development.md:57]: OK

**Comment by yuqi1129** [comment]
@diqiu50 
Please help to look again.

File: docs/trino-connector/development.md (hunk 4 above)
Comment (xiacongling) [reply docs/trino-connector/development.md:27]: yes. i just start gravitino server built from source, and add a test empty metalake. on trino side, there is only `gravitino.properties` under the catalog dir, and Trino dev server started up normally.

Comment (yuqi1129) [reply docs/trino-connector/development.md:27]: I mean if we could list the catalogs in the format like "test.mysql_catalog" by the Trino client.

Comment (yuqi1129) [reply docs/trino-connector/development.md:27]: like 
image

File: docs/assets/trino/pom.xml (hunk 2 above)
Comment (diqiu50) [reply docs/assets/trino/pom.xml]: Using the Maven Central Repository is a good choice. We need to consider situations where a user modifies a class in the client-java-runtime. In such cases, the user might not be able to install the new package.

File: docs/trino-connector/development.md (hunk 6 above)
Comment (diqiu50) [reply docs/trino-connector/development.md]: Ok I understand that. My question is whether anyone else has encountered this error.
@yuqi1129 The steps on how to build Trino are missing.

Comment (yuqi1129) [reply docs/trino-connector/development.md]: @diqiu50 I have added it.
image

File: docs/trino-connector/development.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+## Development environment
+
+To develop the Gravitino connector locally, you need to do the following steps:
+
+### IDEA
+
+1. Clone the Trino repository from the [GitHub](https://github.com/trinodb/trino) repository. We advise you to use the release version 426 or 435. 
+2. Open the Trino project in your IDEA.
+3. Create a new module for the Gravitino connector in the Trino project as the following picture (we will use the name `trino-gravitino` as the module name in the following steps). ![trino-gravitino](../assets/trino/create-gravitino-connector.jpg)
```
Comment (diqiu50) [inline docs/trino-connector/development.md:42]: How to create a new module

Comment (yuqi1129) [reply docs/trino-connector/development.md:42]: This should be tiny things compared to developing The Trino connector. If users lack this knowledge, I don't think they can develop the connector effectively.

File: docs/trino-connector/development.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+The corresponding configuration files are here: [gravitino.properties](../assets/trino/gravitino.properties) and [config.properties](../assets/trino/config.properties).
+
+8. Start the Trino server and connect to the Gravitino server.
+![](../assets/trino/start-trino.jpg)
+9. If `DevelopmentServer` has started successfully, you can connect to the Trino server using the `trino-cli` and run the following command to see if the Gravitino connector is available:
+```shell
+java -jar trino-cli-429-executable.jar --server localhost:8180
+```
+:::note
+The `trino-cli-429-executable.jar` is the Trino CLI jar file, you can download it from the [Trino release page](https://trino.io/download.html).
```
[the remark this answers is not in the document]
Comment (yuqi1129) [reply docs/trino-connector/development.md:75]: It's just an example. It works well for me with version 429, and another version might be just as good, I will add more details and let users choose the package version.

**Comment by yuqi1129** [comment]
@diqiu50 
I have updated the PR, please help take a look if you are free.

**Review by diqiu50 — COMMENTED** [review]
LGTM

**Review by diqiu50 — APPROVED** [review]
